components/supervisor/pkg/supervisor/notification.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package supervisor
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/gitpod-io/gitpod/supervisor/api"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	maxPendingNotifications = 100
+)
+
+// NotificationService implements the notification service API
+type NotificationService struct {
+	mutex                sync.Mutex
+	subscribers          []*api.NotificationService_SubscribeServer
+	nextNotificationId   uint64
+	pendingNotifications map[uint64]*pendingNotification
+}
+
+type pendingNotification struct {
+	message         *api.SubscribeResult
+	responseChannel chan api.NotifyResponse
+}
+
+// RegisterGRPC registers a gRPC service
+func (srv *NotificationService) RegisterGRPC(s *grpc.Server) {
+	api.RegisterNotificationServiceServer(s, srv)
+}
+
+// RegisterREST registers a REST service
+func (srv *NotificationService) RegisterREST(mux *runtime.ServeMux, grpcEndpoint string) error {
+	return api.RegisterNotificationServiceHandlerFromEndpoint(context.Background(), mux, grpcEndpoint, []grpc.DialOption{grpc.WithInsecure()})
+}
+
+// Sends a notification to the user
+func (srv *NotificationService) Notify(ctx context.Context, req *api.NotifyRequest) (*api.NotifyResponse, error) {
+	if srv.nextNotificationId >= maxPendingNotifications {
+		return nil, status.Error(codes.Internal, "Max number of pending notifications exceeded")
+	}
+	srv.mutex.Lock()
```

Comment:
```suggestion
	srv.mutex.Lock()
	defer srv.mutex.Unlock()
```

This way you don't have to make sure that you Unlock the mutex on every return path

Reply:
I don't want to keep the mutex lock when blocking for a result (sounds like a deadlock situation, as the responder tries to get the same lock as well). I guess it's best to factor out the synchronized section into a separate function and use the `defer` approach in that.

Reply:
Correct me if I am wrong, but if we unlock the mutex on return only, we have a blocking wait for the response while holding the lock. The responder will not be able to acquire the mutex thus we have a deadlock.

But using `defer` looks much safer, so I'll extract the locked part of this func into a separate one that uses defer to unlock.